cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -318,28 +324,61 @@ private Future<KafkaRebalance> updateStatus(Reconciliation reconciliation,
         return Future.succeededFuture(kafkaRebalance);
     }
 
-    private RebalanceOptions.RebalanceOptionsBuilder convertRebalanceSpecToRebalanceOptions(KafkaRebalanceSpec kafkaRebalanceSpec) {
-
-        RebalanceOptions.RebalanceOptionsBuilder rebalanceOptionsBuilder = new RebalanceOptions.RebalanceOptionsBuilder();
+    /* test */ AbstractRebalanceOptions.AbstractRebalanceOptionsBuilder<?, ?> convertRebalanceSpecToRebalanceOptions(KafkaRebalanceSpec kafkaRebalanceSpec) {
+
+        AbstractRebalanceOptions.AbstractRebalanceOptionsBuilder<?, ?> rebalanceOptionsBuilder;
+        // backward compatibility, no mode specified means "full-rebalance"
+        KafkaRebalanceMode mode = Optional.ofNullable(kafkaRebalanceSpec)
+                .map(kr -> kr.getMode())
+                .orElse(KafkaRebalanceMode.FULL_REBALANCE);
+        List<Integer> brokers = Optional.ofNullable(kafkaRebalanceSpec)
+                .map(kr -> kr.getBrokers())
+                .orElse(null);
+
+        switch (mode) {
+            case ADD_BROKER:
+                rebalanceOptionsBuilder = new AddBrokerOptions.AddBrokerOptionsBuilder();
+                if (brokers != null && !brokers.isEmpty()) {
+                    ((AddBrokerOptions.AddBrokerOptionsBuilder) rebalanceOptionsBuilder).withBrokers(brokers);
+                } else {
+                    throw new IllegalArgumentException("The brokers list is mandatory when using the add-broker rebalancing mode");
+                }
+                break;
+            case REMOVE_BROKER:
+                rebalanceOptionsBuilder = new RemoveBrokerOptions.RemoveBrokerOptionsBuilder();
```

Comment:
You know if you really want to be fancy you can add the appropriate options builder as a field in the enum class (`REMOVE_BROKER.getOptionsBuilder()`), might even let you generalise this switch statement. Just a thought.

Reply:
That could be cool but anyway doesn't resolve the problem to remove the switch because for different enum values we have to set some different fields. Add/remove broker for now are the same but full rebalance is not (see the default case).